app/Http/Controllers/DemandController.php

```diff
@@ -0,0 +1,334 @@
+<?php
+
+namespace Adshares\Adserver\Http\Controllers;
+
+use Adshares\Esc\Esc;
+
+use Adshares\Adserver\Models\Banner;
+use Adshares\Adserver\Models\Campaign;
+use Adshares\Adserver\Models\EventLog;
+
+use Adshares\Adserver\Http\GzippedStreamedResponse;
+use Adshares\Adserver\Http\Utils;
+
+use Illuminate\Http\Request;
+use Illuminate\Support\Facades\Response;
+
+use Symfony\Component\HttpFoundation\RedirectResponse;
+use Symfony\Component\HttpFoundation\StreamedResponse;
+
+/**
+ * API commands used to serve banners and log relevant events
+ */
+class DemandController extends Controller
+{
+    public function serve(Request $request, $id)
```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Refactor this function to reduce its Cognitive Complexity from 17 to the 15 allowed. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=php%3AS3776)